File: benchmarks/node/04_transactions.js
const Database = require('better-sqlite3');
const path = require('path');

const dbPath = path.resolve(__dirname, '../../databases/bench.db');
const extPath = path.resolve(__dirname, '../../build/plsqlite.so');

function setupAccounts() {
    const db = new Database(dbPath);
    db.prepare("UPDATE accounts SET balance = 1000.0").run();
    db.close();
}

function benchAppLayer(n = 100) {
    const db = new Database(dbPath);
    const start = Date.now();

    const selectStmt = db.prepare("SELECT balance FROM accounts WHERE id = ?");
    const updateFrom = db.prepare("UPDATE accounts SET balance = balance - ? WHERE id = ?");
    const updateTo = db.prepare("UPDATE accounts SET balance = balance + ? WHERE id = ?");

    const transfer = db.transaction((from_id, to_id, amount) => {
        const row = selectStmt.get(from_id);
        if (row && row.balance >= amount) {
            updateFrom.run(amount, from_id);
            updateTo.run(amount, to_id);
        }
    });

    for (let i = 1; i <= n; i++) {
        transfer(i, i + 1, 10.0);
    }

    const elapsed = Date.now() - start;
    db.close();
    return elapsed;
}

function benchPLSQLite(n = 100) {
    const db = new Database(dbPath);
    db.loadExtension(extPath);

    db.prepare(`
        SELECT register_plsql('transfer', 'from_id, to_id, amount', '
            DECLARE bal = 0.0;
            SET bal = (SELECT balance FROM accounts WHERE id = @from_id);
            IF (@bal >= @amount) THEN
                UPDATE accounts SET balance = balance - @amount WHERE id = @from_id;
                UPDATE accounts SET balance = balance + @amount WHERE id = @to_id;
            END IF;
            RETURN @bal;
        ');
    `).run();

    const runStmt = db.prepare("SELECT run_plsql('transfer', ?, ?, 10.0)");

    const start = Date.now();
    for (let i = 1; i <= n; i++) {
        runStmt.get(i, i + 1);
    }

    const elapsed = Date.now() - start;
    db.close();
    return elapsed;
}

function benchPLSQLiteBatched(n = 100) {
    const db = new Database(dbPath);
    db.loadExtension(extPath);

    db.prepare(`
        SELECT register_plsql('batch_transfer', 'n', '
            FOR i IN (WITH RECURSIVE cnt(x) AS (SELECT 1 UNION ALL SELECT x+1 FROM cnt WHERE x < @n) SELECT x FROM cnt) LOOP
                CALL transfer(@i.x, @i.x + 1, 10.0);
            END LOOP;
            RETURN ''DONE'';
        ');
    `).run();

    const start = Date.now();
    db.prepare("SELECT run_plsql('batch_transfer', ?)").get(n);
    const elapsed = Date.now() - start;
    db.close();
    return elapsed;
}

const n = parseInt(process.env.BENCH_TX_COUNT || "10", 10);
console.log(`--- Node.js Transaction Benchmarks (${n} transfers) ---`);

setupAccounts();
const appTime = benchAppLayer(n);
console.log(`RESULT: Transactions: App-Layer: ${appTime}ms`);

setupAccounts();
const plTime = benchPLSQLite(n);
console.log(`RESULT: Transactions: PL/SQLite Proc: ${plTime}ms`);

setupAccounts();
const plBatchTime = benchPLSQLiteBatched(n);
console.log(`RESULT: Transactions: PL/SQLite Batch: ${plBatchTime}ms`);

if (plBatchTime > 0) {
    console.log(`Speedup (App vs Batch): ${(appTime / plBatchTime).toFixed(2)}x`);
}

File: benchmarks/python/04_transactions.py
import sqlite3
import time
import os
import sys

# Path to the extension
if os.name == 'nt':
    EXT_PATH = os.path.abspath("../../build/plsqlite.dll")
else:
    EXT_PATH = os.path.abspath("../../build/plsqlite.so")

DB_PATH = "../../databases/bench.db"

def setup_accounts():
    conn = sqlite3.connect(DB_PATH)
    conn.execute("UPDATE accounts SET balance = 1000.0")
    conn.commit()
    conn.close()

def bench_app_layer(n=1000):
    conn = sqlite3.connect(DB_PATH)
    start = time.time()

    for i in range(1, n + 1):
        from_id = i
        to_id = i + 1
        amount = 10.0

        with conn: # Starts a transaction
            cursor = conn.execute("SELECT balance FROM accounts WHERE id = ?", (from_id,))
            res = cursor.fetchone()
            if res and res[0] >= amount:
                conn.execute("UPDATE accounts SET balance = balance - ? WHERE id = ?", (amount, from_id))
                conn.execute("UPDATE accounts SET balance = balance + ? WHERE id = ?", (amount, to_id))

    end = time.time()
    conn.close()
    return (end - start) * 1000

def bench_plsqlite(n=1000):
    conn = sqlite3.connect(DB_PATH)
    conn.enable_load_extension(True)
    conn.load_extension(EXT_PATH)

    conn.execute("""
    SELECT register_plsql('transfer', 'from_id, to_id, amount', '
        DECLARE bal = 0.0;
        SET bal = (SELECT balance FROM accounts WHERE id = @from_id);
        IF (@bal >= @amount) THEN
            UPDATE accounts SET balance = balance - @amount WHERE id = @from_id;
            UPDATE accounts SET balance = balance + @amount WHERE id = @to_id;
        END IF;
        RETURN @bal;
    ');
    """)

    start = time.time()
    for i in range(1, n + 1):
        conn.execute("SELECT run_plsql('transfer', ?, ?, 10.0)", (i, i + 1))

    end = time.time()
    conn.close()
    return (end - start) * 1000

def bench_plsqlite_batched(n=1000):
    conn = sqlite3.connect(DB_PATH)
    conn.enable_load_extension(True)
    conn.load_extension(EXT_PATH)

    conn.execute("""
    SELECT register_plsql('batch_transfer', 'n', '
        FOR i IN (WITH RECURSIVE cnt(x) AS (SELECT 1 UNION ALL SELECT x+1 FROM cnt WHERE x < @n) SELECT x FROM cnt) LOOP
            CALL transfer(@i.x, @i.x + 1, 10.0);
        END LOOP;
        RETURN "DONE";
    ');
    """)

    start = time.time()
    conn.execute("SELECT run_plsql('batch_transfer', ?)", (n,))
    end = time.time()
    conn.close()
    return (end - start) * 1000

if __name__ == "__main__":
    n = int(os.getenv("BENCH_TX_COUNT", "10"))
    print(f"--- Python Transaction Benchmarks ({n} transfers) ---")

    setup_accounts()
    app_time = bench_app_layer(n)
    print(f"RESULT: Transactions: App-Layer: {app_time:.2f}ms")

    setup_accounts()
    pl_time = bench_plsqlite(n)
    print(f"RESULT: Transactions: PL/SQLite Proc: {pl_time:.2f}ms")

    setup_accounts()
    pl_batch_time = bench_plsqlite_batched(n)
    print(f"RESULT: Transactions: PL/SQLite Batch: {pl_batch_time:.2f}ms")

    print(f"Speedup (App vs Proc): {app_time/pl_time:.2f}x")
    print(f"Speedup (App vs Batch): {app_time/pl_batch_time:.2f}x")
